[ServiceBus] Update migration guide (#15915)
Azure/azure-sdk-for-python: sdk/servicebus/azure-servicebus/migration_guide.md

You can create senders and receivers from this client to the queue/topic/subscription/session of your
choice and start sending/receiving messages.

#### Approachability
By having a single entry point, the `ServiceBusClient` helps with the discoverability of the API
as you can explore all available features through methods from a single client, as opposed to searching
through documentation or exploring namespace for the types that you can instantiate.
Whether sending or receiving, using sessions or not, you will start your applications by constructing the same client.

#### Consistency
We now have methods with similar names, signature and location to create senders and receivers.
This provides consistency and predictability on the various features of the library.

### Client constructors

- While we continue to support connection strings when constructing a client, the main difference is when using Azure Active Directory.
We now use the new [`azure-identity`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md) library
to share a single authentication solution between clients of different Azure services.

> In V0.50:

Reviewer: Not blocking this PR, but thoughts for a future PR

We want to show 3 different things here
- That constructing the client using connection string has not changed
- Constructing client using SAS key and name is an upcoming feature
- Using AAD is a brand new feature

Consider treating these separately. For example:

> Constructing the client using the connection string is same in both versions.
`servicebus_client = ServiceBusClient.from_connection_string(conn_str)`

> Additionally, you can now use Azure Active Directory for authentication via the `azure-identity` library
`show code for AAD here`

> The option to construct the client using SAS key name and value is not available in the new version, but will be added in the near future.

Author: I love the way you organize the concepts, I have copied your words and tweaked a bit.

- The `send` method is renamed to `send_messages` following the pattern of using the `messages` suffix in methods that deal with messages.
- You can now pass an array of messages directly in method used to send messages rather than first creating a `BatchMessage` 
- Sending multiple messages in a single go always had the potential to fail if batch size exceeded the size limit. To help with this, we have a new class `ServiceBusMessageBatch` which helps in creating a batch that will never increase the size limit.

In V0.50:
```python
with sender:
    # send a single message
    sender.send(Message("Hello world!"))

    # send multiple messages. This can fail if the batch exceeded size limit
    sender.send(BatchMessage(["data 1", "data 2", ...]))
```

In V7:
```python
with sender:

    # send a single message
    sender.send_messages(ServiceBusMessage("Hello world!"))

    # send multiple messages. This can fail if the batch exceeded size limit
    sender.send_messages([ServiceBusMessage("data1"), [ServiceBusMessage("data2"), [ServiceBusMessage("data3")])

    # safely send multiple messages by using a batch object
    message_batch = sender.create_message_batch()

>     message_batch.add_message(ServiceBusMessage("data"))

Author: comment added

  - `label` has been renamed to `subject`
  - `user_properties` has been renamed to `application_properties`
  - `annotations` has been placed under `raw_amqp_message` instance variable.
- `ServiceBusReceivedMessage` is now the class representing the message when you get it from the service, regardless of whether you used the `peek_messages`/`receive_deferred_messages` operation or received it using the receiver.
  - Properties `settled` and `expired` are no longer available.
- Refer to the [CHANGELOG.md](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/servicebus/azure-servicebus/CHANGELOG.md) for more changes on the message properties.

### Working with Administration Client

In v0.50, you could create/get/update/delete/list Service Bus queues/topics/subscriptions/rules using the `control_client`.
In v7, this is replaced by the `ServiceBusAdministrationClient`.
The following code snippets show how to manage queues, similar methods are provided on the `ServiceBusAdministrationClient` to manage topics, subscriptions and rules.

> #### Managing Queues

Reviewer: This header is not needed anymore

```suggestion
```

## Table of contents

* [Migration benefits](#migration-benefits)
    - [Cross Service SDK improvements](#cross-service-sdk-improvements)
* [Important changes](#important-changes)
    - [Client hierarchy](#client-hierarchy)
    - [Client constructors](#client-constructors)
    - [Creating sender and receivers](#creating-sender-and-receivers)
    - [Sending messages](#sending-messages)
    - [Receiving messages](#receiving-messages)
    - [Settling messages](#settling-messages)
    - [Renewing lock on the received message](#renewing-lock-on-the-received-message)
    - [Scheduling messages and cancelling scheduled messages](#scheduling-messages-and-cancelling-scheduled-messages)
    - [Working with sessions](#working-with-sessions)
    - [Working with UTC time](#working-with-utc-time)
    - [Working with AutoLockRenewer](#working-with-autolockrenewer)
    - [Working with Message properties](#working-with-message-properties)
    - [Working with Administration Client](#working-with-administration-client)
    - [Migration samples](#migration-samples)

> * [Addition samples](#addition-samples)

Reviewer 2: ```suggestion
* [Additional samples](#additional-samples)
```

By having a single entry point, the `ServiceBusClient` helps with the discoverability of the API
as you can explore all available features through methods from a single client, as opposed to searching
through documentation or exploring namespace for the types that you can instantiate.
Whether sending or receiving, using sessions or not, you will start your applications by constructing the same client.

#### Consistency
We now have methods with similar names, signature and location to create senders and receivers.
This provides consistency and predictability on the various features of the library.

### Client constructors

- While we continue to support connection strings when constructing a client, the main difference is when using Azure Active Directory.
We now use the new [`azure-identity`](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md) library
to share a single authentication solution between clients of different Azure services.
- `QueueClient`, `TopicClient` and `SubscriptionClient` have been removed, to create senders and receivers, please check [Creating sender and receivers](#creating-sender-and-receivers)

> In V0.50:

Reviewer 2: What is the reason the table format is changed? Also, I think there's a great value in providing the relevant sample in place

```python
servicebus_client = ServiceBusClient.from_connection_string()

> servicebus_client = ServiceBusClient()

Author: AAD is not supported in 0.50, we could mention that in the `New Feature` section.

queue_client = QueueClient.from_connection_string()
topic_client = QueueClient.from_connection_string()
subscription_client = QueueClient.from_connection_string()
```

In V7:
```python
# Authenticate with connection string
servicebus_client = ServiceBusClient.from_connection_string()

# Authenticate with Active Directory:
from azure.identity import DefaultAzureCredential
servicebus_client = ServiceBusClient(fully_qualified_namespace, credential=DefaultAzureCredential())
```

### Creating sender and receivers

- `QueueClient`, `TopicClient` and `SubscriptionClient` have been replaced with methods to create receivers and senders directly from `ServiceBusClient`.

> - To create a receiver for dead-letter sub queue, `get_deadletter_receiver` has been replaced with `get_<queue/subscritpion>_receiver` which takes an optional parameter `sub_queue` for indication.

Reviewer: ```suggestion
- The method to create a receiver takes an optional parameter `sub_queue` which can be used to create a receiver for dead-letter sub queue. This replaces the dedicated method `get_deadletter_receiver` from the older version.
```

> In V0.50:

Reviewer 2: What is the reason the table format is changed? Also, I think there's a great value in providing the relevant sample in place

Author: I totally agree with you that table has the advantage for comparison.

but what I found is that table might be good for comparison of short/brief code snippets, but too crowded for large content.

for example this one:
![image](https://user-images.githubusercontent.com/47871814/103578550-9f011100-4e8b-11eb-9b8d-a1227a607bde.png)

I think the readability decreases as we want to show more contents in a table. one approach might be we further compress the code snippets, but flattening it out also help users who just want to copy and past the code.

---

actually, I think `Message properties migration guide segment` is a perfect case to apply table.

```python
client = ServiceBusClient.from_connection_string(connstr)

> queue_client = client.get_queue(queue_name)
> queue_sender = queue_client.get_sender()
> queue_receiver = queue_client.get_receiver()
> queue_dead_letter_receiver = queue_client.get_deadletter_receiver()

Reviewer: ```suggestion
// for queues
queue_client = client.get_queue(queue_name)
queue_sender = queue_client.get_sender()
queue_receiver = queue_client.get_receiver()
queue_dead_letter_receiver = queue_client.get_deadletter_receiver()
```

> topic_client = client.get_topic(topic_name)
> topic_sender = topic_client.get_sender()
> 
> subscription_client = client.get_subscription(topic_name, subscription_name)
> subscription_receiver = subscription_client.get_receiver()
> subscription_dead_letter_receiver = subscription_client.get_deadletter_receiver()
> ```

Reviewer: ```suggestion
// for topics
topic_client = client.get_topic(topic_name)
topic_sender = topic_client.get_sender()

// for subscriptions
subscription_client = client.get_subscription(topic_name, subscription_name)
subscription_receiver = subscription_client.get_receiver()
subscription_dead_letter_receiver = subscription_client.get_deadletter_receiver()
```

In V7:
```Python

> with ServiceBusClient.from_connection_string(connstr) as client:

Reviewer: Any reason to use the `with` construct in the v7 code snippet but not in v0.50?

Author: it's because v5 `ServiceBusClient` doesn't support the `with` context manager.
but I tend to keep the usage of `with` for v7 for consistency with other snippets.

    queue_sender = client.get_queue_sender(queue_name)
    queue_receiver = client.get_queue_receiver(queue_name)
    queue_dead_letter_receiver = client.get_queue_receiver(queue_name, sub_queue=ServiceBusSubQueue.DEAD_LETTER)

    topic_sender = client.get_topic_sender(topic_name)

    subscription_receiver = client.get_subscription_receiver(topic_name, subscription_name)
    subscription_dead_letter_receiver = client.get_subscription_receiver(topic_name, subscription_name, sub_queue=ServiceBusSubQueue.DEAD_LETTER)
```

### Sending messages

> - `send` method is renamed to `send_messages` on `ServiceBusSender` to be consistent in usage of the `messages` suffix in other methods on the receiver and the sender.
> - `BatchMessage` is replaced with `ServiceBusMessageBatch` to send a batch of messages in one call not exceeding the size limit of the sender.

Reviewer: The main thing to notice for the user is not that there has been a rename from `send` to `send_messages`. I believe that would be straight forward for them to notice. The main catch is that if they want to send an array of messages, previously they need to construct the `BatchMessage`, but now the same send method that takes a single message will also take the array of messages directly.

> `BatchMessage` is replaced with `ServiceBusMessageBatch` to send a batch of messages 

That is not accurate... `BatchMessage` is simply removed. Previously one needed to create it by passing an array of messages, now they can pass the array of messages directly to the send method. `ServiceBusMessageBatch` is an entirely new concept..

```suggestion
- The `send` method is renamed to `send_messages` following the pattern of using the `messages` suffix in methods that deal with messages.
- You can now pass an array of messages directly in method used to send messages rather than first creating a `BatchMessage` 
- Sending multiple messages in a single go always had the potential to fail if batch size exceeded the size limit. To help with this, we have a new class `ServiceBusMessageBatch` which helps in creating a batch that will never increase the size limit.
```

In V0.50:
```python
with sender:

>     sender.send(Message("Hello world!"))
>     sender.send(BatchMessage(["data 1", "data 2", ...]))

Reviewer: ```suggestion
    // Send a single message
    sender.send(Message("Hello world!"))
    
    // Send multiple messages. This can fail if the batch exceeded size limit
    sender.send(BatchMessage(["data 1", "data 2", ...]))
```

```

In V7:
```python
with sender:

>     sender.send_messages(ServiceBusMessage("Hello world!"))

Reviewer: ```suggestion
    // Send a single message
    sender.send_messages(ServiceBusMessage("Hello world!"))
```

>     message_batch = sender.create_message_batch()
>     message_batch.add_message(ServiceBusMessage("data"))
>     sender.send_messages(message_batch)

Reviewer: ```suggestion
    // Send multiple messages. This can fail if the batch exceeded size limit
    sender.send_messages(["data 1", "data 2", ...])
    
    // Safely send multiple messages by using a batch object
    message_batch = sender.create_message_batch()
    message_batch.add_message(ServiceBusMessage("data"))
    sender.send_messages(message_batch)
```

    peeked_messages = receiver.peek_messages(max_message_count=1, sequence_number=None)
```

### Settling messages

- Previously, the methods to settle messages (`complete()`, `abandon()`, `defer()` and `dead_letter()`) were on the messages themselves.
These have been moved to the receiver in the new version, take the message as input and have the `message` suffix in their name.
- The idea is to have the message represents just the data and not have the responsibility of any operation on the service side.
Additionally, since a message cannot be settled if the receiver that was used to receive it is not alive, tying these operations
to the receiver drives the message home better.

In V0.50:
```python

with receiver:
    received_message.complete()
    # or
    # received_message.abandon()
    # received_message.defer()
    # received_message.dead_letter()
```

In V7:
```python
with receiver:
    receiver.complete_message(received_message)
    # or
    # receiver.abandon_message(received_message)
    # receiver.defer_message(received_message)
    # receiver.dead_letter_message(received_message)
```

> ### Renewing lock on the received message

Reviewer: Can we combine the auto lock renewal section in here? Since both topics are about lock renewals it would make sense to have a single section called `Lock renewal`

Author: good suggestion, I tried to combine them into one, let me know how it looks to you.

- `renew_lock` method was moved from `Message` to `ServiceBusReceiver` and renamed to `renew_message_lock` for the same reason as settlement methods movement.

In V0.50:
```python
with receiver:
    received_message.renew_lock()
```

In V7:
```python
with receiver:
    receiver.renew_message_lock(received_message)
```

> ### Scheduling messages and cancelling scheduled messages

Reviewer: There is not much in this section other than renames. In which case I would either recommend not calling them out at all or having a separate section called "Renames" (or something nicer :) ) and put these there. Having a dedicated section is helpful if the method has moved or has changed in what it accepts as input or changes what it returns as output.

Other sections that fall under the same category are the ones for utc time and for receiving messages

Author: There is a change in the schedule/cancel schedule api that in v0.50 the method takes `argument list` while in v7 it takes `list` so I'd like to keep it.

I have removed the utc time section; as for the `receiving messages` sections since it's a champion scenario I tend to keep it as it is.

- `shutdown` method is renamed to `close` to keep consistent with sender and receiver classes.

In V0.50:
```python
from azure.servicebus import AutoLockRenew
auto_lock_renew = AutoLockRenew()

with receiver:
    auto_lock_renew.register(received_message)

with session_receiver:
    auto_lock_renew.register(session_receiver)

auto_lock_renew.shutdown()
```

In V7:
```python
from azure.servicebus import AutoLockRenewer
auto_lock_renewer = AutoLockRenewer()

with receiver:
    auto_lock_renew.register(receiver, received_message)

with session_receiver:
    auto_lock_renew.register(session_receiver, session_receiver.session)

auto_lock_renew.close()
```

### Working with Message properties

There're properties on `ServiceBusMessage` being added/removed/changed, please check the table follow:

| In v0.50 | Equivalent in v7 | Note |
|---|---|---|
| `azure.servicebus.Message.user_properties` | `azure.servicebus.ServiceBusMessage.application_properties` | Some message properties have been renamed, e.g. accessing the application specific properties of a message. |

### Working with Administration Client

> The management related operations like creating, updating, deleting and listing queues, topics, subscriptions or rules are to be done with the `ServiceBusAdministrationClient` which can be imported from the `azure.servicebus.management`. If one was relying on `control_client` available in v0.50, and cannot migrate to the `azure-mgmt-servicebus` package (or requires SAS based authentication for management), one would utilize this migration segment, otherwise, the aforementioned dedicated mgmt package should be used.

Reviewer: We don't have any reasons why we should be telling customers who are using this package to do the management operations to move to the mgmt package. So, we don't need to call out `azure-mgmt-servicebus` here.

```suggestion
In v0.50, you could create/get/update/delete/list Service Bus queues/topics/subscriptions/rules using the `control_client`. In v7, this is replaced by the `ServiceBusAdministrationClient`.
```

#### Managing Queues

In V0.50:
```python
from azure.servicebus.control_client import ServiceBusService
service_bus_service = ServiceBusService()
queue = service_bus_service.get_queue(queue_name)
service_bus_service.create_queue(queue_name)
service_bus_service.delete_queue(queue_name)
queues = service_bus_service.list_queues()
```

In V7:
```python
from azure.servicebus.management import ServiceBusAdministrationClient
service_bus_administration_client = ServiceBusAdministrationClient()
queue = service_bus_administration_client.get_queue(queue_name)
service_bus_administration_client.create_queue(queue_name)
service_bus_administration_client.delete_queue(queue_name)
queues = service_bus_administration_client.list_queues()
```

> #### Managing Topics

Reviewer: I would recommend removing the sections for topics, subscriptions and rules. Instead, we can add a line saying "You have similar methods on the ServiceBusAdministrationClient to manage topics, subscriptions and rules"
